nmigen_boards/atlys.py

```diff
@@ -0,0 +1,240 @@
+            Subsignal("scl",     Pins("D9"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("sda",     Pins("C9"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("clk",     DiffPairs("B6", "A6", dir="o")),
+            Subsignal("d",       DiffPairs("D8 C7 B8", "C8 A7 A8", dir="o")),
+            Attrs(IOSTANDARD="TMDS_33"),
+        ),
+        Resource("hdmi", 2, # J3, input only due to on board buffer, HDMI A connector
+            Subsignal("scl",     Pins("M16"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("sda",     Pins("M18"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("clk",     DiffPairs("H17", "H18", dir="i")),
+            Subsignal("d",       DiffPairs("K17 L17 J16", "K18 L18 J18", dir="i")),
+            Attrs(IOSTANDARD="TMDS_33"),
+        ),
+        Resource("hdmi", 3, # JA, input/output as it is unbuffered, HDMI D connector
+            Subsignal("scl",     Pins("C13"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("sda",     Pins("A13"), Attrs(IOSTANDARD="I2C")),
+            Subsignal("clk",     DiffPairs("T9", "V9")),
+            Subsignal("d",       DiffPairs("R3 T4 N5", "T3 V4 P6")),
+            Attrs(IOSTANDARD="TMDS_33"),
+        ),
+
+        Resource("ac97", 0,
+            Subsignal("clk",     Pins("L13", dir="o")),
+            Subsignal("sync",    Pins("U17", dir="o")),
+            Subsignal("reset",   Pins("T17", dir="o")),
+            Subsignal("sdo",     Pins("N18", dir="o")),
+            Subsignal("sdi",     Pins("T18", dir="i")),
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
+    ]
+    connectors = [
+        Connector("pmod", 0, "T3 R3 P6 N5 - - V9 T9 V4 T4 - -"), # JB
+
+        Connector("vhdci", 0, # JC
+            "U16 - U15 U13 - M11 R11 - T12 N10 - M10 U11 - R10 - - - - U10 - R8  M8  - U8  U7  - N7  T6  - R7  N6  - U5 "
+            "V16 - V15 V13 - N11 T11 - V12 P11 - N9  V11 - T10 - - - - V10 - T8  N8  - V8  V7  - P8  V6  - T7  P7  - V5 "
+        ),
+    ]
+
+    def __init__(self, JP12="2V5"):
+        assert(JP12 in ["2V5", "3V3"])
+
+        super().__init__()
+
+        # Voltage on bank2 is configured by JP12
+        iostd = "LVCMOS25" if JP12 == "2V5" else "LVCMOS33"
+        self.add_resources([
+            Resource("user_led", 7, Pins("N12",  dir="o"), Attrs(IOSTANDARD=iostd)), # LD7
+
+            Resource("user_btn", 0, PinsN("T15", dir="i"), Attrs(IOSTANDARD=iostd)), # RESET
+            Resource("reset"   , 0, PinsN("T15", dir="i"), Attrs(IOSTANDARD=iostd)), # RESET
+
+            Resource("user_sw" , 4, Pins("P12",  dir="i"), Attrs(IOSTANDARD=iostd)), # SW4
+            Resource("user_sw" , 5, Pins("R5",   dir="i"), Attrs(IOSTANDARD=iostd)), # SW5
+            Resource("user_sw" , 6, Pins("T5",   dir="i"), Attrs(IOSTANDARD=iostd)), # SW6
+        ])
+
+    def toolchain_program(self, products, name):
+        subprocess.run(["impact", "-batch", "{}.impact".format(name)],
+            cwd=products._root, check=True,
```

Comment:
You should never access internal (starting with `_`) fields of `BuildProducts` (or anything else in nMigen) directly.

This is a problem here because while the current `BuildProducts` implementation provides all the files locally, I would like to add one in the future that performs the entire build remotely, such that `products.get()` performs a transfer over SSH. See the Versa ECP5 for an example of how to do it correctly.

Reply:
Difficulty was that the generated file `{{name}}.impact` needs to refer to other generated file `{{name}}.bit`. I now switched to generate the commands inside `toolchain_program()`

Reply:
I see, thanks. Another solution might be to guarantee that each file grabbed via `extract` ends up in the same directory. But it might not help enough.

Reply:
And I also needed to guarantee that the programming command was executed in the directory where the files are extracted.